[Compose 페이먼츠 2단계] 공백 미션 제출합니다

junseo511 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
- 안녕하세요, 동전! 레아의 폭풍 리뷰를 받다보니 코드를 거의 처음부터 작성하게 되었네요 😓
- 자세한 이야기는 PR 내부의 `file changes` 에 달아두도록 하겠습니다 :)

## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

https://github.com/user-attachments/assets/30aca639-ac00-4b69-8990-18f8b9c2d3b2

junseo511 committed:
docs: README 업데이트

junseo511 committed:
feat: 카드 번호 입력 필드 개선 및 테스트 코드 추가

junseo511 committed:
feat: 카드 만료일 입력 필드 개선 및 테스트 코드 추가

junseo511 committed:
feat: 카드 소유자 이름 입력 필드 개선 및 테스트 코드 추가

junseo511 committed:
feat: 비밀번호 입력 필드 개선 및 테스트 코드 추가

junseo511 committed:
refactor: 카드 추가 화면 입력 필드 리팩터링 및 UI 테스트 코드 제거

junseo511 committed:
feat: 카드 지갑 화면 추가 및 시작 화면으로 설정

- `CardWalletActivity`를 추가하고, 앱의 시작 화면으로 설정했습니다.
- 기존의 `NewCardActivity`는 더 이상 시작 화면이 아닙니다.

junseo511 committed:
feat: 카드 UI 모델 추가

- `CardUiModel` 데이터 클래스를 추가했습니다.
- `kotlin-parcelize` 플러그인을 추가했습니다.

junseo511 committed:
feat: 카드 지갑 화면 UI 구현

- `CardWalletScreen` UI를 구현했습니다.
- 카드 목록 상태에 따라 다른 UI를 보여주도록 `CardWalletState`를 추가하고, `CardWalletContent`에서 분기 처리했습니다.
- `PaymentCard` 컴포저블을 수정하여 카드 정보를 표시하도록 했습니다.
- 카드 추가 화면(`NewCardActivity`)에서 카드를 추가하면 결과를 반환하고, `CardWalletScreen`에서 이를 처리하여 카드 목록에 추가하도록 구현했습니다.

junseo511 committed:
refactor: Preview에 누락된 테마 적용

junseo511 committed:
test: 카드 지갑 화면 UI 테스트 코드 작성

junseo511 committed:
refactor: 코드 포맷팅

junseo511 committed:
refactor: 공통으로 사용되는 `PaymentCard` 컴포저블 및 `CardUiModel` 위치 변경

junseo511 committed:
refactor: `CARD_NUMBER_SEPARATOR` 상수화 및 중복 제거

junseo511 committed:
docs: 2단계 기능 구현 완료 표시

junseo511 commented on app/src/androidTest/java/woowacourse/payments/ui/cardwallet/CardWalletContentTest.kt:
```diff
@@ -0,0 +1,86 @@
+import woowacourse.payments.ui.cardwallet.model.CardWalletState
+import woowacourse.payments.ui.common.model.CardUiModel
+
+class CardWalletContentTest {
```
Screen을 테스트하는 과정에서 레아의 'Controller` 를 테스트 하는 것과 무엇이 다른가? 라는 질문을 받아 각 컴포넌트별 테스트를 분리했습니다

junseo511 commented on app/src/main/java/woowacourse/payments/ui/cardwallet/CardWalletScreen.kt:
```diff
@@ -0,0 +1,100 @@
+    val state = CardWalletState.from(cards.size)
+
+    val launcher =
+        rememberLauncherForActivityResult(
```
과제 요구사항에 맞게 해당 Screen 내에서 액티비티 런처를 사용해 진행했습니다!

junseo511 commented on app/src/main/java/woowacourse/payments/ui/cardwallet/components/CardWalletContent.kt:
```diff
@@ -0,0 +1,66 @@
+    state: CardWalletState,
+    navigateToNewCard: () -> Unit,
+) {
+    when (state) {
```
각 장수별 State를 정의하고, 상태에 따른 UI를 출력하도록 진행했습니다!

junseo511 commented on app/src/main/java/woowacourse/payments/ui/transformation/GroupedOffsetMapping.kt:
```diff
@@ -0,0 +1,32 @@
+import androidx.compose.ui.text.AnnotatedString
+import androidx.compose.ui.text.input.OffsetMapping
+
+class GroupedOffsetMapping(
```
텍스트 변환에 대한 가독성을 개선하고자 기존의 함수 형태와 달리 별도의 객체로 분리했습니다! 또한, value값을 직접적으로 바꾸지 않으면서 안정성을 높였습니다 :)

junseo511 commented on app/src/main/java/woowacourse/payments/ui/common/model/components/PaymentCard.kt:
```diff
@@ -0,0 +1,139 @@
+}
+
+@Composable
+private fun CardContents(
```
```
안녕하세요, 여러분~~ 오랜만에 잡담-질답에 글을 올려보네요
안드 크루원분들은 컴포즈에서 개발을 진행할 때 Composable 함수를 어떻게 분리하시나요?

- 한 kt 파일에 Composable 함수가 두 개 이상 존재. (예시: Card 컴포넌트 파일 안에 CardContent)
- 다른 kt 안에 별도의 Composable 함수로 존재 (예시: CardWalletScreen 밖에 별도의 NewCardButton 분리)

재사용이 기준일까요? 아니면 한 번 사용되더라도 가독성 개선이나 분리된 유지보수 등을 위해 분리?
어떤 기준으로 파일을 분리하고 통합해야 하는지 여러분의 의견이 궁금합니다 🤔
```

[슬랙에 남긴 글](https://woowacourse.slack.com/archives/C08BEPG1SSE/p1757403569073109)이었는데 이에 대한 동전의 의견도 궁금하네요!
저같은 경우에는 해당 CardContents는 절대적으로 PaymentCard 내부에서만 사용되고,
귀속되는 정보라 생각해 이곳에 위치하도록 두었습니다 :)

donghyun81 replied:
이번에 UiModel과 UiState를 비교하면서 생각해봤는데, 저는 composable 함수를 포함해서 상위 범위에서 공통적인 책임과 역할을 가지고 있다면 상위 범위로 추출할 것 같아요.
예를 들어 CardUiModel은 카드를 Ui에서 표시하는 책임과 역할을 가지고 있어서 Ui 레이어에 model 패키지에 두고
CreateCardUiState는 한 화면 범위의 상태를 관리하는 책임과 역할을 가지고 있어서 각 화면 패키지에 model 패키지에 둔다고 생각했어요.
CardContents라는 컴포넌트가 PaymentCard뿐만 아니라 다른 컴포넌트에도 같은 책임과 역할을 가진다고 생각할때 추출하면 될 것 같다고 생각해요~

woowahan-leah replied:
UIModel과 UIState를 꼭 분리할 필요가 있었을까요? (뭐라 하려는거 아님, 단순 질문)

donghyun81 replied:
책임과 역할이 달라서 분리할 필요가 있다고 생각해요.
UiModel이든 UiState든 명칭이 중요한건 아니지만, UiModel은 Ui레이어에서는 다 같은 model에 책임과 역할을 해야하고 UiState는 각 화면에 상태에서만 같은 역할과 책임을 가지고 있는 것 같아요.
결국 어디 범위까지 같은 책임과 역할을 가지냐에 따라서 더 넓은 범위에서 사용될 수 있다는 생각을 가지고 있어요 그래서 분리할 필요가 있다고 느꼈습니다!

junseo511 replied:
저는 개인적으로 UI Model과 UI State를 분리하거나 묶는 것은 자유라고 생각합니다만,
분리되는 것이 가독성이나 관리 측면에서 더 용이하다고 생각되었습니다.
그리고 동전의 말처럼 UI State는 해당 화면만을 담당하고 UI Model이 여러 곳에서 사용되는 것이라면,
일관성을 위해서라도 분리하는 것이 좋아보입니다 :)

donghyun81 commented on app/src/androidTest/java/woowacourse/payments/ui/cardwallet/CardWalletContentTest.kt:
```diff
@@ -0,0 +1,86 @@
+        composeRule
+            .onNodeWithText("+")
+            .assertIsDisplayed()
+    }
```
State는 SINGLE인데 카드가 복수형인 것 같아요! 비슷한 테스트 내용이 많아서 전체적으로 확인 해주시면 좋을 것 같아요~

junseo511 replied:
직접 입력하기보다 해당 state는 enum class의 팩토리 메서드를 활용해볼 수 있을 것 같네요 🤔
해당 작업은 state holder가 진행해주고 있으니 :)
[refactor: CardWalletState 생성 로직 변경](https://github.com/woowacourse/android-payments/pull/40/commits/e16339c4177422c44bf57b09b35a22c5b68ee163)

donghyun81 commented on app/src/androidTest/java/woowacourse/payments/ui/newcard/CardHolderTextFieldTest.kt:
```diff
@@ -0,0 +1,104 @@
+        // then
+        composeRule
+            .onNodeWithText("공백")
+            .assertIsDisplayed()
```
문자와 공백만 입력 가능하다는 테스트명인데 텍스트가 입력 되었을때 그대로 나오는지 테스트 인 것 같다고 느껴집니다!
문자와 공백을 제외한 내용 표시가 안되는 테스트 내용이 있으면 좋을 것 같아요~

junseo511 replied:
정말로 그렇네요! 살짝 바꿔서 반영하겠습니다 :)
확인 감사합니다 👍
[refactor: 카드 소유자 이름 테스트 케이스 수정](https://github.com/woowacourse/android-payments/pull/40/commits/0ed55d0c91dec78a32f3e0d90101dd9dba35d89c)

donghyun81 commented on app/src/androidTest/java/woowacourse/payments/ui/newcard/CardNumberTextFieldTest.kt:
```diff
@@ -0,0 +1,158 @@
+        // then
+        composeRule
+            .onNodeWithText("1234123412341234")
+            .assertIsDisplayed()
```
16자를 입력하고 16자를 테스트하고 있어요. 입력한 데이터가 표시된다는 테스트와 같다고 느껴져서 16자 제한을 위한 테스트라면 아래 테스트와 중복되는 것 같아요!

donghyun81 commented on app/src/main/java/woowacourse/payments/ui/cardwallet/model/CardWalletState.kt:
```diff
@@ -0,0 +1,17 @@
+                cardCount <= 0 -> EMPTY
+                cardCount == 1 -> SINGLE
+                else -> MULTIPLE
+            }
```
enum class로 CardWallet의 상태를 관리하면 각 상태 별로 어떤 속성을 가져야 하는지 분기하기 어렵다고 생각해요.
테스트 코드에서 single과 multiple의 테스트가 잘못 진행된 것도 이 부분에서 발생할 것 같은데 어떻게 생각하시는지 궁금합니다!

junseo511 replied:
enum class보다 명시적으로 '상수'를 표현할 수 있는 방법이 있을까요? Holder가 충분히 이를 보완해줄 수 있다고 생각됩니다 :) 테스트 코드에서는 해당 팩토리 메서드 패턴을 적극 활용하는 것도 방법이 될 수 있겠어요!

donghyun81 replied:
제가 생각했던 부분은 Empty,Single,Multiple이 특정 상태에 가깝다고 생각했어요!  
enum class로 상태를 상수를 표현하기보다 sealed interface로 카드 목록을 어떤 상태인지로 분리하면 좋을 것 같다고 생각했어요!
이렇게 생각한 이유는 각자 필요한 속성이 다르다고 느껴서인데,
Empty 상태는 Card가 없고, Single 상태는 Card를 가지고, Multiple은 상태는 Cards를 가지고 있어야 한다고 생각해요.
해당 이점은 enum class와 Cards 목록을 분리해서 유지 보수하지 않아도 된다는 이점이 있는 것 같아요.
Cards 리스트에 카드가 추가되거나 삭제된다면 Cards 목록에 따라 enum class의 상태를 변경해야 하는데,
enum class를 수정하지 않아도 컴파일에서 알 수가 없어요!
sealed interface를 사용한다면 추가되거나 삭제될 때 각 상태에 맞는 속성을 강제로 할당해야 해서 컴파일 과정에서 확인하기 좋다고 생각이었어요!
해당 생각은 블랙잭에서 보유한 카드 개수에 따라 다른 상태가 되었던 기억과 프로젝트에서 화면이 Loadind,Success,Fail일때 상태를 기준으로 화면을 그리는 것이 유지보수에 좋았던 것 같아서 이야기해봤어요!

donghyun81 commented on app/src/main/java/woowacourse/payments/ui/common/model/CardUiModel.kt:
```diff
@@ -0,0 +1,11 @@
+    val numberDigits: String,
+    val expiry: String,
+    val holder: String,
+) : Parcelable
```
password는 제외 해주신 것 좋아요!

donghyun81 commented on app/src/main/java/woowacourse/payments/ui/common/model/components/PaymentCard.kt:
```diff
@@ -0,0 +1,139 @@
+                    color = White,
+                    fontSize = 12.sp,
+                    fontWeight = FontWeight.W500,
+                ),
```
TextStyle이 카드의 정보들을 표시할때 공통적인 스타일이라고 생각하면 추출해도 좋을 것 같고, 변경하는 이유가 각자 달라야 한다면 그대로 유지하면 좋을 것 같다고 생각하는데 공백은 어떻게 생각하시나요??

MyStoryG replied:
공통 스타일이라면 스타일만 추출하면 좋을까요, Text 컴포저블 자체를 재사용하는 것이 좋을까요? 공백은 어떻게 생각하시나요??

junseo511 replied:
> TextStyle이 카드의 정보들을 표시할때 공통적인 스타일이라고 생각하면 추출해도 좋을 것 같고, 변경하는 이유가 각자 달라야 한다면 그대로 유지하면 좋을 것 같다고 생각하는데 공백은 어떻게 생각하시나요??

디자인 시스템이 규정되지 않은 상황에서 공통 시스템을 추출하는 것에 큰 의미가 있을까요? 공통 디자인을 정리하면서 생길 수 있는 장단점에는 어떤 것들이 있을까요?

junseo511 replied:
> 공통 스타일이라면 스타일만 추출하면 좋을까요, Text 컴포저블 자체를 재사용하는 것이 좋을까요? 공백은 어떻게 생각하시나요??
@MyStoryG 

공통 스타일은 style만 사용하는 것이 좋을 것 같습니다!
만약 각 Text 컴포저블에 일부는 굵은 글씨로 써야 하고 일부는 색상이 들어가야 하는 등 기능이 추가되어 하나씩 늘리기 시작한다면,
오히려 원하는 컴포넌트를 찾기 어렵고 해당 컴포저블 함수는 가지고 있는 책임이 늘어날 것 같아요 :)
그게 아니라면 불필요한 뎁스가 발생할 것 같구요!
스타일을 분리해 사용하는 것이 보다 유연한 방식이 될 것 같다고 생각합니다 🤔

donghyun81 replied:
이번에 프로젝트 하면서 정말 많이 느낀 부분인데 공통 디자인 정리는 함부로 하는 게 아니라고 많이 느꼈어요.
각 컴포넌트간에 공통 내용이 겹치는 부분도 아닌 부분도 있어서 오히려 혼란만 엄청나게 많이 생기는 부분이 많았어요.
그래서 공통 스타일 부분의 위치를 해당 PaymentCard Composable에서만 하는 건 어떨까 하는 제안이었습니다!
제가 디자인을 잘 몰라서 공통 디자인을 처음부터 잘 설계하기는 어려운 것 같아서 작은 단위의 하위 Composable부터 공통 스타일을 추출해보면서 더 큰 단위로 추출해보는 방식은 어떨까 해서 제안했었어요~

donghyun81 commented on app/src/main/java/woowacourse/payments/ui/newcard/NewCardScreen.kt:
```diff
@@ -59,90 +51,88 @@ fun NewCardScreen(
+                value = cardNumber,
+                onValueChange = { cardNumber = it },
+                modifier = Modifier.fillMaxWidth(),
+                onImeAction = { focusManager.moveFocus(FocusDirection.Next) },
```
onImeAction에서 포커스를 다음으로 이동시켜주고 있는데 KeyBoard 각 컴포넌트들의 KeyboardOption은 ImeAction.Next로 정해져 있는 것 같아요. 하위 컴포넌트는 상위에서 onImeAction이 무엇을 하는지 모르는데 하위에서 ImeAction.Next로 고정하는 것이 좋을지 공백 의견이 궁금합니다~

junseo511 replied:
하위 컴포넌트가 알아야 할 이유가 무엇인지 궁금합니다! 상위에서 만약에 적어야 하는 칸이 하나가 더 생긴다면,
하위 컴포넌트는 `상위 컴포넌트에 입력해야 하는 칸이 추가되었어!` 를 알고 스스로 바뀌어야 할까요?
관리하는 위치는 포커스가 직접 이동시킬 수 있는 상위 객체에서 진행되어야 한다고 생각합니다 :)

donghyun81 replied:
제가 생각한 것도 공백과 일치해요 그래서 하위 컴포넌트가 KeyBoardOption에 ImeAction도 같이 몰라야 하지 않은가 하는 의견이었습니다!
자세하게는 상위에서 KeyBoard 옵션에 onImeAction과 ImeAction을 같이 내려야 할 것 같다는 생각입니다!
해당 구현에서 상위 컴포넌트에서 순서가 변경되거나 추가 되었을 때 Pin 번호의 onImeAction은 다음 포커스로 이동으로 변경했는데,  
PinTextField의 ImeAction은 하위 컴포넌트에서 고정되어 있어서 상위의 순서가 변경될 때 하위가 수정되어야 하는 구현이 될 것 같아요!

donghyun81 commented on app/src/main/java/woowacourse/payments/ui/cardwallet/components/CardWalletContent.kt:
```diff
@@ -0,0 +1,66 @@
+@Composable
+fun CardWalletContent(
+    cards: List<CardUiModel>,
+    state: CardWalletState,
```
이렇게 별도의 파라미터로 분리하면 cards는 개수가 없는데 state는 SINGLE 상태로 인자가 들어올 수 있을 것 같아요!
사용하는 곳에서 항상 cards랑 state의 상태가 알맞게 들어온다고 확신 할 수 있을까요??

junseo511 replied:
> 이렇게 별도의 파라미터로 분리하면 cards는 개수가 없는데 state는 SINGLE 상태로 인자가 들어올 수 있을 것 같아요! 사용하는 곳에서 항상 cards랑 state의 상태가 알맞게 들어온다고 확신 할 수 있을까요??

이제 해당 작업을 수업 시간에 배운 State Holder가 진행하고, 외부에서는 그저 holder의 값을 주도록 리팩토링을 진행해보았습니다 :)
[refactor: CardWalletState 관련 로직 CardWalletUiStateHolder로 이전](https://github.com/woowacourse/android-payments/pull/40/commits/53d9381d8cb22172a1815a26f16bd031d2093f02)

donghyun81 commented on app/src/main/java/woowacourse/payments/ui/cardwallet/components/CardWalletContent.kt:
```diff
@@ -0,0 +1,66 @@
+
+        CardWalletState.MULTIPLE -> {
+            CardWalletCards(cards)
+        }
```
각 상태에 따른 독립된 화면이라고 생각하는데, 여기서 각각 그려도 괜찮을까요??
그리고 화면에 그려진 상태가 상위에서 무조건 Column에서 그려진다는 확신이 있는 것 같아요!

junseo511 replied:
충분히 그럴 수 있겠네요! 쿨 반영 하겠습니다 :)
[refactor: CardWalletContent 컴포저블 분리](https://github.com/woowacourse/android-payments/pull/40/commits/ffb74054216ff0cde26a49bd790e267cb4e77e10)

donghyun81 reviewed (changes requested):
저번에 컴포즈를 이해하는데 혼란을 많이 느꼈습니다. 
스스로 확신이 없었던 부분이 많아서 리뷰를 많이 못 했던 것 같아서 레아가 직접 리뷰를 꼼꼼하게 잘 남겨주신 것 같아요!
이전 리뷰 반영을 위해서 변경 점이 많았을 텐데 구현 잘해주신 것 같고 테스트도 많이 해주신 것이 역시 공백이라고 느껴집니다! 질문 남겨주신 것 포함해서 이번에는 제 생각을 이전보다는 작성해서 이야기를 많이 해볼 수 있을 것 같아요~ 
제 의견이 추상적일 수도 있어서 궁금한 부분 메시지 보내주시면 최대한 답장해 보겠습니다~

junseo511 committed:
refactor: `CardWalletState` 관련 로직 `CardWalletUiStateHolder`로 이전

junseo511 committed:
refactor: transformation 패키지 이동

junseo511 committed:
refactor: CardWalletState 생성 로직 변경

junseo511 committed:
refactor: PaymentCard 컴포저블 경로 변경

junseo511 committed:
refactor: `NewCardScreen` 상태 관리 로직 분리

junseo511 committed:
refactor: 공백 관련 테스트 케이스 추가

junseo511 committed:
refactor: state 생성 안정성 강화

junseo511 committed:
refactor: 상태 홀더 적용에 따른 파라미터 이름 변경

junseo511 committed:
fix: 카드 정보 입력 UI 로직 수정

- 카드 정보 입력 시 최대 글자 수를 넘지 않도록 제한
- 카드 소유자 이름 입력 시 영문/숫자만 입력되도록 수정
- 카드 개수 제한 2개 -> 1개로 변경

junseo511 committed:
refactor: 카드 소유자 이름 테스트 케이스 수정

junseo511 reviewed (commented):
바쁘신데 리뷰 감사합니다 동전!
저도 현생 이슈로 지금 리뷰 요청을 보내네요 :)
끝까지 파이팅해봅시다 함께...ㅎ

junseo511 committed:
refactor: CardWalletContent 컴포저블 분리

donghyun81 commented on app/src/main/java/woowacourse/payments/ui/newcard/model/NewCardUiStateHolder.kt:
```diff
@@ -0,0 +1,107 @@
+    expiryInit: String = "",
+    holderInit: String = "",
+    pinInit: String = "",
+    scrollInit: Int = 0,
```
스크롤 상태까지는 생각 못해봤네요! 잘 배워가겠슴돠.
단순 궁금증인데 스크롤 상태를 관리할 수 있으면 포커스도 상태로 관리하는 것이 좋을까요??

donghyun81 reviewed (approved):
공백 구현을 볼때마다 느끼는건 요구사항에 표시되지 않는 사용자가 경험할 부분을 상세하게 신경쓰는 부분이 많아 보여요!
다양한 프로젝트 경험과 공백의 세심한 부분이 포함되어 있는 것 같아서 저도 신경 써야겠다고 생각하는 부분을 많이 배워가는 것 같습니다!
공백의 의견 남겨주신 부분에서 답장을 남겼고 3단계에서 더 이야기 해보면 좋을 것 같아요~
